hledger-lib/Hledger/Data/DayPartition.hs

```diff
@@ -0,0 +1,269 @@
+splitSpan _      (MonthAndDay m d)        ds = splitspan (nextmonthandday m d)   addGregorianYearsClip              1 ds
+splitSpan _      (DaysOfWeek [])          _  = Nothing
+splitSpan _      (DaysOfWeek days@(n:_))  ds = do
+    (s, e) <- dateSpanSplitLimits (nthdayofweekcontaining n) nextday ds
+    let -- can't show this when debugging, it'll hang:
+        bdrys = concatMap (\d -> map (addDays d) starts) [0,7..]
+        -- The first representative of each weekday
+        starts = map (\d -> addDays (toInteger $ d - n) $ nthdayofweekcontaining n s) days
+    spansFromBoundaries e bdrys
+
+
+-- | Fill in missing start/end dates for calculating 'splitSpan'.
+dateSpanSplitLimits :: (Day -> Day) -> (Day -> Day) -> DateSpan -> Maybe (Day, Day)
+dateSpanSplitLimits _     _    (DateSpan Nothing   Nothing) = Nothing
+dateSpanSplitLimits _     _    ds | isEmptySpan ds          = Nothing
+dateSpanSplitLimits start _    (DateSpan (Just s) (Just e)) = Just (start $ fromEFDay s, fromEFDay e)
+dateSpanSplitLimits start next (DateSpan (Just s) Nothing)  = Just (start $ fromEFDay s, next $ start $ fromEFDay s)
+dateSpanSplitLimits start next (DateSpan Nothing  (Just e)) = Just (start $ fromEFDay e, next $ start $ fromEFDay e)
+
+-- Split the given span into exact spans using the provided helper functions:
+--
+-- 1. The start function is used to adjust the provided span's start date to get the first sub-span's start date.
+--
+-- 2. The next function is used to calculate subsequent sub-spans' start dates, possibly with stride increased by a multiplier.
+--    It should handle spans of varying length, eg when splitting on "every 31st of month",
+--    it adjusts to 28/29/30 in short months but returns to 31 in the long months.
+splitspan :: (Day -> Day) -> (Integer -> Day -> Day) -> Int -> DateSpan -> Maybe DayPartition
+splitspan start next mult ds = do
+    (s, e) <- dateSpanSplitLimits start (next (toInteger mult)) ds
+    let bdrys = mapM (next . toInteger) [0,mult..] $ start s
+    spansFromBoundaries e bdrys
+
+-- | Construct a list of exact 'DateSpan's from a list of boundaries, which fit within a given range.
+spansFromBoundaries :: Day -> [Day] -> Maybe DayPartition
+spansFromBoundaries _ []     = Nothing
+spansFromBoundaries e (x:_)  | x >= e = Nothing
+spansFromBoundaries e xs     = boundariesToMaybeDayPartition $ takeUntilFails (<e) xs
+  where
+    takeUntilFails _ []     = []
+    takeUntilFails p (y:ys) = if not (p y) then [y] else y : takeUntilFails p ys
+
+
+-- Get the natural start for the given interval that falls on or before the given day,
+-- when applicable. Works for Weeks, Months, Quarters, Years, eg.
+intervalBoundaryBefore :: Interval -> Day -> Day
+intervalBoundaryBefore i d =
+  case dayPartitionToNonEmpty <$> splitSpan True i (DateSpan (Just $ Exact d) (Just . Exact $ addDays 1 d)) of
+    Just ((start, _) :| _ ) -> start
+    _ -> d
+
+
+intToDay = ModifiedJulianDay . toInteger
```

Comment:
> The Modified Julian Day is a standard count of days, with zero being the day 1858-11-17.

This sounds like it might not fit with our support of all years ? (CE currently, maybe BCE some day)

Reply:
Yes, this is from an earlier commit, and yes it technically does not support all years.

- On a 64-bit system this will support all years between 252 quadrillion BC (well before the development of currency) and 252 quadrillion AD (at which point we'll be more worried about powering our computers since all stars will have become black holes or white dwarfs).
- On a 32-bit system, it will support all years between 5 877 751 BC and 5 881 468 AD.

We could extend support to all integers. This would incur a small performance penalty as we'd have to switch from IntMap to Map internally. I personally think the supported date ranges are fine, but happy for you to make a call.

Reply:
Oh wow. So 0 to 1858 is not a problem then. Great.